### ISSUE TITLE
Support possibility to enable arbitrary language extensions for HLint 

### ISSUE THREAD

**Issue by geraldus** [flycheck/flycheck#638]
HLint does not enables `TemplateHaskell` and `QuasiQuotes` language extensions, so if you have quasi quotes in your code HLint fails with parse error. The only way to enable mentioned extensions is to create auxiliary script like following

```sh
#!bin/sh
hlint -XQuasiQuotes -XTemplateHaskell -hGeneralise -hDefault -hDollar "$@"
```

flag it as executable and set it as FlyCheck HLint Executable. Though it works as expected it would be nice to have an option to configure list of additional extensions inside Emacs.

P.S. Maybe there is some way to configure extension used by HLint per project like hint files (e.g. HLint.hs is project root), but I was not able to figure out that.

**Comment by lunaryorn** [comment]
That's just a matter of adding the proper options.  I'll see whether I can do that by the week end.

As for your PS: I don't know.  It'd only be possible if hlint itself supported such files.  You'll need to read the hlint documentation for more information.

**Comment by vlatkoB** [comment]
Would be nice to change it so it looks for `HLint.hs` file in the project root directory, which is the standard, IIRC.
Just tested this code, and it works OK.
```
  :command ("hlint" "-XQuasiQuotes" "-XTemplateHaskell"
    (eval
     (let ((hlinths
            (concat (flycheck-module-root-directory (flycheck-find-in-buffer flycheck-haskell-module-re))
                    "../HLint.hs")))
     (if (file-exists-p hlinths)
        (concat "-h" hlinths))))
    source-inplace)
```
Or to run it from project root, so `hlint` can pick up `HLint.hs` by itself.

As per extensions, `ghc-mod` has `ghc-hlint-options` for hlint args, so maybe something similarly named.

(Also maybe change `haskell-ghc` checker to `:command ("ghc" "-Wall" "-no-link" "-outputdir" "/tmp/flycheck-haskell"` -[wedens](https://github.com/wedens/flycheck/commit/ce3e325a5803fd5cd31ad18a87475f18e18cd00b), so no more manual editing after every update)

I'm using elisp and emacs just a few weeks, so please do optimize the code. :-)

**Comment by lunaryorn** [comment]
> Would be nice to change it so it looks for HLint.hs file in the project root directory, which is the standard, IIRC.

What does `Hlint.hs` do?   If it is some kind of configuration file for Hlint, then please change your code to use the infrastructure that Flycheck provides for this purpose.  See `python-pylint` for an example.

Besides, a pull request would help.  I'm fine with `haskell-hlint` as it is now, and while I see the need for options for language extensions and will add them eventually, I do have more important and more interesting things to do currently.

> Or to run it from project root, so hlint can pick up HLint.hs by itself.

Flycheck does not support that, see #312.

> As per extensions, ghc-mod has ghc-hlint-options for hlint args, so maybe something similarly named.

I'm not sure what you're trying to tell me.  Flycheck doesn't use `ghc-mod`.

> (Also maybe change haskell-ghc checker to :command ("ghc" "-Wall" "-no-link" "-outputdir" "/tmp/flycheck-haskell" -wedens, so no more manual editing after every update)

Again, I'm not sure what you mean.   I don't know why I should be concerned with your “manual editing”, but if you'd like me to make a certain change to Flycheck, I do expect a better reason and explanation.

**Comment by vlatkoB** [comment]
Yes `HLint.hs` is a configuration file for hlint, and it is placed in project root. It can be specific for each project. It is the file that you specify what to report/ignore. Small example:
```
import "hint" HLint.Default
ignore "Use camelCase" = ""
ignore "Use mappend" = ""
```
I'll take a look at `python-pylint`. Thanks for the tip.


It was just a proposal for the name of var that would allow proper customization of `hlint` in `flycheck`. I think geraldus was thinking about something like that.


Without that manual editing, `flycheck` [doesn't work] (https://github.com/flycheck/flycheck-haskell/issues/6).

**Comment by geraldus** [comment]
@vlatkoB hi!  I still does not know is it possible to enable extensions per project for HLint via `HLint.hs`, maybe you know? 

> It was just a proposal for the name of var that would allow proper customization of hlint in flycheck. I think geraldus was thinking about something like that.

Yes, it is.  But tweaking a bit default `haskell-ghc` seems better solution for me (I'm still unfamiliar with most GHC's options and can't be certain about flags used in **wedens** version).  Also, in some cases Flycheck does not report error messages, but I still have no minimal example for this.  Anyway, the final word is for Sebastian.

**Comment by vlatkoB** [comment]
@geraldus Think extensions can't be specified by `HLint.hs`. 
I made some simple code that seems to work correctly. It allows customizing any `hlint` arg, so maybe it might be possible to specify options locally via `projectile` or `.dir-locals.el`.
```
(flycheck-def-option-var flycheck-hlint-options
    '("-XQuasiQuotes" "-XTemplateHaskell") haskell-hlint
  "A list of additional hlint options.

The value of this variable is a list of strings, where each
string is full hlint option, with its switch.

Refer to http://community.haskell.org/~ndm/darcs/hlint/hlint.htm for
more information."
  :type '(repeat :tag "Options" (string :tag "Option"))
  :safe #'flycheck-string-list-p
  :package-version '(flycheck . "0.20"))

(flycheck-define-checker haskell-hlint
  "A Haskell style checker using hlint.

See URL `https://github.com/ndmitchell/hlint'."
  :command ("hlint"
            (config-file "-h" flycheck-hlintrc)
            (option-list "" flycheck-hlint-options concat)
            source-inplace)
  :error-patterns
  ((warning line-start
            (file-name) ":" line ":" column
            ": Warning: "
            (message (one-or-more not-newline)
                     (one-or-more "\n" (one-or-more not-newline)))
            line-end)
   (error line-start
          (file-name) ":" line ":" column
          ": Error: "
          (message (one-or-more not-newline)
                   (one-or-more "\n" (one-or-more not-newline)))
          line-end))
  :modes (haskell-mode literate-haskell-mode))
```
This is just a draft, but you can try it by yourself.

 @lunaryorn Would something like this be acceptable?

**Comment by geraldus** [comment]
Cool!

**Comment by vlatkoB** [comment]
@geraldus 
>I still does not know is it possible to enable extensions per project for HLint via >HLint.hs, maybe you know?

It would be possible now with `.dir-local.el`.  
Just set `flycheck-hlint-language-extensions` to something like ` '("QuasiQuotes" "TemplateHaskell")`

**Comment by geraldus** [comment]
Great news!
